File: c/src/Creational/Builder/boston_soupbuffet_builder.c

//BostonSoupBuffetBuilder - One of Two Builder Subclasses

#include "../../common/bostonclamchowder.h"
#include "../../common/bostonfishchowder.h"

#include "soupbuffet_builder.h"
#include "boston_soupbuffet_builder.h"


soupbuffet_builder_t * boston_soupbuffet_builder_new()
{
	soupbuffet_builder_t * sbb = soupbuffet_builder_new( boston_soupbuffet_builder_init );
	return sbb;
}

void boston_soupbuffet_builder_init(soupbuffet_builder_t * sbb)
{
	sbb->build_clamchowder = boston_soupbuffet_builder_buildClamChowder;
	sbb->build_fishchowder = boston_soupbuffet_builder_buildFishChowder;
	sbb->set_soupbuffet_name = boston_soupbuffet_builder_setSoupBuffetName;
}

void boston_soupbuffet_builder_buildClamChowder(soupbuffet_builder_t * sbb)
{
	sbb->soupbuffet->clamChowder = bostonclamchowder_new();
}

void boston_soupbuffet_builder_buildFishChowder(soupbuffet_builder_t * sbb) 
{
	sbb->soupbuffet->fishChowder = bostonfishchowder_new();
}    
    
void boston_soupbuffet_builder_setSoupBuffetName(soupbuffet_builder_t * sbb) 
{
	sbb->soupbuffet->name = "Boston Soup Buffet";
}

File: ruby/src/Creational/Builder/BostonSoupBuffetBuilder.rb

#/BostonSoupBuffetBuilder - Two of Two Builder Subclasses

require "Creational/Builder/SoupBuffetBuilder"
require "common/BostonClamChowder"
require "common/BostonFishChowder"

class BostonSoupBuffetBuilder < SoupBuffetBuilder

	def initialize
	end

	def buildClamChowder()
   		@soupBuffet.clamChowder = BostonClamChowder.new
	end

	def buildFishChowder()
   		@soupBuffet.fishChowder = BostonFishChowder.new
	end

	def setSoupBuffetName()
 		@soupBuffet.setSoupBuffetName("Boston Soup Buffet")
	end

end
